n3fit/src/n3fit/model_trainer.py

```diff
@@ -349,12 +363,21 @@ def _model_generation(self, pdf_models, partition, partition_idx):
         # The input layer was a concatenation of all experiments
         # the output of the pdf on input_layer will be thus a concatenation
         # we need now to split the output on a different array per experiment
+
         sp_ar = [self.input_sizes]
         sp_kw = {"axis": 1}
+        splitting_op = op.as_layer(op.split, op_args=sp_ar, op_kwargs=sp_kw, name="pdf_split")
         splitting_layer = op.as_layer(op.split, op_args=sp_ar, op_kwargs=sp_kw, name="pdf_split")
-        splitted_pdf = splitting_layer(full_pdf_per_replica)
+        # splitted_pdf = splitting_layer(full_pdf_per_replica)
+
+        splitted_pdfs = []
+        for layer in all_replicas_pdf:
+            split_layers = splitting_op(layer)
+            splitted_pdfs.append([op.expand(sl) for sl in split_layers])
```

Comment:
I'm confused about this changes. The way it is done now `full_pdf_per_replica` and `splitting_layer` are not needed.

The original logic was to have just one (very big) pdf that would convolute with the fktable for the many replica case. Have you tested that the changes doesn't lead to tf doing many convolutions? (one per replica) We would want to avoid that (and it should have no effect on having a trvl different per replica).

If you have (and it doesn't) then please remove `splitting_layer` and `full_pdf_per_replica` (since this is actually cleaner than what I had anyway).

Reply:
Yes those are remnants of earlier code and will be cleaned.

I don't think tf is doing more convolutions per se in this new architecture, but it probably cannot do them as efficiently as in the earlier version, where all replica tensors were essentially stacked no?

Reply:
Well, it is a difference between doing (pseudo code)

```python
pdf_{ri} X fktable_{i} = result_{r}
```

or

```python
result = []
for i in pdf_list:
    result.append(pdf_{i} X fktable_{i})
```
Specially in a GPU, the first version will be much faster than the second as there is only one copy of the PDF for all replicas (even if it is a bigger array) instead of many copy of smaller PDFs.

Reply:
In your second example, the allocation for the convolution will probably be the bottleneck on the GPU. But then again, tensorflow uses a big memory pool so this isn't an issue. As for the extra copies: you're right, this architecture has more layers, more tensors being copied and probably less utilization of the device. But we will measure this to avoid speculation.

As I said, this was exploratory from my side and I am going to focus my attention on getting the tr-vl split into the masked convolutions at the level of the Observable implementations. Currently the masks are only there to denote active flavors (uniform per replica of course); the aim will be to mask out data points per replica, right?

Reply:
No, there are two masks. The flavour mask is applied to the pdf/luminosity and it is always the same for all replicas for a given fktable.

Instead, the training/validation mask is different per replica. Currently it is applied on the fktable since it reduces its size considerably. However, by doing that, a different trvl mask generates a new fktable, which is a bottleneck for parallelizarion over the number of replicas.

Instead the trvl mask would need to be applied to the data, so that all replicas use the same fktable and generate the same (in shape and operations) output vector and only at the end the validation is data is masked out

Reply:
Ok so where the observable class now contains a list of fk-tables (I assume length n-replicas, each entry a masked version of the full fk-table), this will be only the full one. When you say 'at the end the validation data is masked out' you mean the observable layer should perform an extra masking of the fk-table convolution. So for instance DY will first execute a masked convolution, then a tensor product with the fk-tables (uniform across replicas) and finally masks out the validation entries per replica.

Reply:
The fitting process is as follows (for a DY-type observable which is the most complicated one). The exact order of the shape might not be correct in the example below though. 

x enters as an input to the PDF model, input shape (n_x,)

The PDF model is a concatenation of PDFs, one for replica, its output is then (nreplicas, n_flavours, n_x)

The result of the PDF models is then given to the observable. The observable can be formed by several fktables (for instance, the observable can be the measure of a ratio between two cross sections).
Inside the observable first a luminosity tensor is constructed to account for the two colliding protons, so that we have a luminosity such that:

L = PDF x PDF, shape (nreplicas, n_flavours, n_flavours, n_x, n_x)

However, not all flavour combinations contribute to all observables, so this luminosity tensor is masked such that only active flavours remain (which means the calculation to be done is also lighter).

The new masked luminosity is then: mL = (nreplicas, n_active_flavours_combinations, n_x, n_x)

Now this masked luminosity is convoluted with every fktable in the observable. The output of this convolution is a vector of data points (n_data).

If the observable is composed (for instance, a ratio between two cross sections or whatever) then an operation is applied to all the output arrays.

The final result of the observable is then (n_replicas, n_data).

-----------------------------------------------------------------------------------------------------------

However, in order to do the fit correctly we apply a training/validation mask. This traning validation mask is applied at the level of the data points, so the natural way would be, after the computation is finished, apply the mask, so that one gets:

training_observable = observable X mask
validation_observable = observable X ~mask

But, for backpropagation, we are only interested in the training_observable, and the fktable is a massive object. Therefore what we do is to apply the mask directly to the fktable so that we have

fk_training = fk X mask
fk_validation = fk X ~mask

And then, using those two fktables, we generate two observables (so we no longer have to apply the mask at the end).

Note btw that this training/validation mask to the fktable is applied before the training even starts.

Reply:
ok thanks for your explanation @scarlehoff, I was a bit confused by the list of fk-tables within the observable class, but your comment clarifies this.